go/ql/src/experimental/CWE-20/MemoryAllocationDos.ql

```diff
@@ -0,0 +1,73 @@
+/**
+ * @name User-controlled data control memory allocation size
+ * @description Using a user input value to memory allocation size
+ * @kind path-problem
+ * @problem.severity error
+ * @id go/allow-memory-size-injection
+ * @problem.severity 7.1
+ * @tag security
+ *      experimental
+ *      external/cwe/cwe-023
+ */
+
+import go
+import semmle.go.security.FlowSources
+import DataFlow
+import semmle.go.security.AllocationSizeOverflow
+import DataFlow::PathGraph
+
+module MemoryAllocationDos {
+  abstract class Source extends DataFlow::Node { }
+
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  // If developer comparison size, it should be safe
+  class CompSanitizer extends Sanitizer {
```

Comment:
Implement this as a sanitizer guard: for example,

```
private predicate redirectCheckGuard(DataFlow::Node g, Expr e, boolean outcome) {
  e = g.(DataFlow::RelationalComparisonNode).getAnOperand().asExpr() and
  outcome = [true, false]
}

...

override predicate isSanitizer(DataFlow::Node node) {
  node = DataFlow::BarrierGuard<redirectCheckGuard/3>::getABarrierNode()
}
```

Reply:
I tried this before, but found that it only blocks the taint inside the `if `block. What did I miss?